File: leetcode/problems/language-specific/go/695.max-area-of-island.go
/*
 * @lc app=leetcode id=695 lang=golang
 *
 * [695] Max Area of Island
 */

// @lc code=start

// * Recursive DFS Solution | O(rc) Time | O(rc) Space
// * r -> Total number of rows in a grid matrix | c -> Total number of cols in a grid matrix

func maxAreaOfIsland(grid [][]int) int {
	ROWS, COLS := len(grid), len(grid[0])
	maxArea, seen := 0, make([][]bool, ROWS)
	for i := 0; i < ROWS; i++ {
		seen[i] = make([]bool, COLS)
	}

	var maxAreaOfIslandHelper func(row, col int) int
	maxAreaOfIslandHelper = func(row, col int) int {
		if row < 0 || row > ROWS-1 ||
			col < 0 || col > COLS-1 ||
			seen[row][col] || grid[row][col] == 0 {
			return 0
		}

		seen[row][col] = true
		return 1 + (maxAreaOfIslandHelper(row-1, col) + maxAreaOfIslandHelper(row+1, col) + maxAreaOfIslandHelper(row, col-1) + maxAreaOfIslandHelper(row, col+1))
	}

	for row := 0; row < ROWS; row++ {
		for col := 0; col < COLS; col++ {
			maxArea = findMax(maxArea, maxAreaOfIslandHelper(row, col))
		}
	}

	return maxArea
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

// @lc code=end

File: leetcode/problems/sessions/2023-05-01_2023-??-??/695.max-area-of-island.py
#
# @lc app=leetcode id=695 lang=python3
#
# [695] Max Area of Island
#

# @lc code=start

# * Recursive DFS Solution | O(mn) Time | O(1) Space
# * m -> The number of rows in the grid matrix | n -> The number of columns in the grid matrix


class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        ROWS, COLS = len(grid), len(grid[0])

        def dfs(row, col):
            if (
                not 0 <= row <= ROWS - 1
                or not 0 <= col <= COLS - 1
                or grid[row][col] == 0
            ):
                return 0

            grid[row][col] = 0
            cur_area = 0
            for d_row, d_col in DIRECTIONS:
                cur_area += dfs(row + d_row, col + d_col)

            return 1 + cur_area

        max_area = 0
        for row in range(ROWS):
            for col in range(COLS):
                max_area = max(max_area, dfs(row, col))

        return max_area


# @lc code=end
